orttraining/orttraining/core/optimizer/triton_fusion.cc

```diff
@@ -0,0 +1,432 @@
+using NodeVec = InlinedVector<Node*>;
+using NodeArgVec = InlinedVector<NodeArg*>;
+using ConstNodeArgVec = InlinedVector<const NodeArg*>;
+using NodeArgSet = InlinedHashSet<NodeArg*>;
+using IsSupportedFunc = std::function<bool(const Graph&, const Node&)>;
+
+int64_t Hash(const std::string& str) {
+  uint32_t hash = 0;
+  for (char const& c : str) {
+    hash = hash * 101 + c;
+  }
+
+  return static_cast<int64_t>(hash);
+}
+
+bool CheckAxis(const Node& node, int64_t expected_axis) {
+  const auto& attributes = node.GetAttributes();
+  if (attributes.find("axis") == attributes.end() || !attributes.at("axis").has_i()) return false;
+  int64_t axis = attributes.at("axis").i();
+  if (axis == expected_axis) return true;
+  if (axis < 0 || expected_axis < 0) {
+    const auto& input_shape = node.InputDefs()[0]->Shape();
+    if (input_shape == nullptr) return false;
+    int64_t rank = static_cast<int64_t>(input_shape->dim_size());
+    if (axis < 0) axis += rank;
+    int64_t non_neg_expected_axis = expected_axis < 0 ? expected_axis + rank : expected_axis;
+    return axis == non_neg_expected_axis;
+  }
+  return false;
+}
+
+bool CheckAxes(const Graph& graph, const Node& node, bool single_axis, const std::vector<int>& expected_axes) {
+  std::vector<int64_t> axes_values;
+  const auto& attributes = node.GetAttributes();
+  if (attributes.find("axes") != attributes.end()) {
+    axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+  } else if (node.InputDefs().size() == 2) {
+    auto axes_const = graph.GetConstantInitializer(node.InputDefs()[1]->Name(), true);
+    if (!axes_const) {
+      return false;
+    }
+    Initializer initializer{*axes_const, graph.ModelPath()};
+    axes_values.insert(axes_values.end(), initializer.DataAsSpan<int64_t>().begin(),
+                       initializer.DataAsSpan<int64_t>().end());
+  } else {
+    return false;
+  }
+
+  if (expected_axes.empty()) {
+    return !single_axis || axes_values.size() == 1;
+  }
+
+  std::vector<int64_t> expected_axes_values(expected_axes.begin(), expected_axes.end());
+  bool has_negative_axis = false;
+  for (auto axis : axes_values) {
+    if (axis < 0) {
+      has_negative_axis = true;
+      break;
+    }
+  }
+  if (!has_negative_axis) {
+    for (auto axis : expected_axes_values) {
+      if (axis < 0) {
+        has_negative_axis = true;
+        break;
+      }
+    }
+  }
+  if (has_negative_axis) {
+    const auto& input_shape = node.InputDefs()[0]->Shape();
+    if (input_shape == nullptr) return false;
+    int64_t rank = static_cast<int64_t>(input_shape->dim_size());
+    for (auto& axis : axes_values) {
+      if (axis < 0) axis += rank;
+    }
+    for (auto& axis : expected_axes_values) {
+      if (axis < 0) axis += rank;
+    }
+  }
+  std::sort(axes_values.begin(), axes_values.end());
+  std::sort(expected_axes_values.begin(), expected_axes_values.end());
+  return axes_values == expected_axes_values;
+}
+
+struct Partition {
```

Comment:
It's better to have comments here

Reply:
What are supporting nodes? Maybe change to: A partition contains all connected nodes in the ONNX graph which are supported by Triton?

Reply:
Shall we rename this to TritonPartition? or TritonOpPartition? IMO Partition is an overloaded term in ORT and making the name more intuitive will reduce confusion.